File: apps/platform/ditto/api_models/submission_deposit_address.py
"""Audited operator control for the miner submission deposit address."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated

from pydantic import AfterValidator, BaseModel, ConfigDict, Field
from scalecodec.utils.ss58 import ss58_decode

_SS58_PATTERN = r"^[1-9A-HJ-NP-Za-km-z]{32,64}$"


def _valid_bittensor_ss58(address: str) -> str:
    try:
        ss58_decode(address, valid_ss58_format=42)
    except (TypeError, ValueError) as error:
        raise ValueError("must be a valid Bittensor SS58 address") from error
    return address


SubmissionPaymentAddress = Annotated[
    str,
    Field(pattern=_SS58_PATTERN),
    AfterValidator(_valid_bittensor_ss58),
]


class SubmissionDepositAddressRevision(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    revision: int
    parent_revision: int
    payment_address: Annotated[str, Field(pattern=_SS58_PATTERN)]
    reason: str
    actor: str
    created_at: datetime | None


class AdminSubmissionDepositAddressResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    current: SubmissionDepositAddressRevision
    history: list[SubmissionDepositAddressRevision]


class AdminSubmissionDepositAddressRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True, str_strip_whitespace=True)

    expected_revision: Annotated[int, Field(ge=0)]
    payment_address: SubmissionPaymentAddress
    reason: Annotated[str, Field(min_length=8)]
    actor: Annotated[str, Field(min_length=1, max_length=120)] = "admin_api"
    confirmation: str

File: apps/backroom/src/lib/submission-deposit-address.ts
import { z } from 'zod'

export const SS58_ADDRESS_PATTERN = /^[1-9A-HJ-NP-Za-km-z]{32,64}$/

export const submissionDepositAddressRevisionSchema = z.object({
  revision: z.number().int().nonnegative(),
  parent_revision: z.number().int().nonnegative(),
  payment_address: z.string().regex(SS58_ADDRESS_PATTERN),
  reason: z.string(),
  actor: z.string(),
  created_at: z.string().datetime({ offset: true }).nullable(),
})

export const submissionDepositAddressControlSchema = z.object({
  current: submissionDepositAddressRevisionSchema,
  history: z.array(submissionDepositAddressRevisionSchema).max(100),
})

export const updateSubmissionDepositAddressInputSchema = z.object({
  expectedRevision: z.number().int().nonnegative(),
  paymentAddress: z.string().trim().regex(SS58_ADDRESS_PATTERN),
  reason: z.string().trim().min(8),
  confirmation: z.string(),
})

export function submissionDepositAddressConfirmation(address: string) {
  return `SET SUBMISSION DEPOSIT ADDRESS ${address}`
}

export type SubmissionDepositAddressControl = z.infer<
  typeof submissionDepositAddressControlSchema
>
